docs/backend/test.md

```diff
@@ -0,0 +1,105 @@
+> make pythontests_watch        # adds a watcher that can run pytest in the background; unit tests will re-run whenever changes to a Python file are detected
+> make pythontests_coverage     # run pytest with coverage report
+> make pythontests ARGS='registration/tests/<file_name.py>' # run pytest for a specific file
+> make pythontests ARGS='-k <TestClassname>' # run pytest for a specific class, e.g. make pythontests ARGS='-k TestNaicsCodeEndpoint'
+> make pythontests ARGS='-k <test_name>' # run pytest for a specific test, e.g. make pythontests ARGS='-k test_get_method_for_200_status' (note: if any tests have the same name, even if they're within different classes, this command will run them all)
+```
+
+## Testing Helpers
+
+We have some testing helpers in the `tests/utils` directory of each Django app. These helpers are used to mock data and objects for testing purposes. For example, in the `registration` app, we have a helper called `TestUtils` that contains the following methods:
+
+- mock user roles for get, post, and put requests
+- mock postal codes
+- authorize a user as belonging to an operator
+- create mock operations
+
+To use the helpers, import them from `utils` and use like this:
+
+```python
+TestUtils.mock_postal_code()
+```
+
+## Writing backend tests using mock
+
+We use the `mock` library (a.k.a. `unittest.mock`) to patch calls made in our backend. For examples of this in our codebase, search for the keyphrase `mocker.patch`.
+
+When mocking calls, it's very important to insert your patch in the correct place - namely, patch where an object is _called_ (looked up), not where it is defined. For a more detailed explanation, see [Where to patch](https://docs.python.org/3/library/unittest.mock.html#id6).
+
+It's also important to understand the difference between patching a function - done with `patch()` - and patching an object - done with `patch.object()`. Again, refer to unittest.mock's [online docs](https://docs.python.org/3/library/unittest.mock.html#patch) for more detail.
+
+## Testing Django Models
+
+When testing Django models, we use the `BaseTestCase` class from the `bc_obps.common.tests.utils.helpers` module. This base test case provides several utility methods to streamline the testing of model fields, ensuring that they meet the expected criteria for labels, maximum lengths, and relationship counts. Below are the key components and their purposes:
+
+### Key Components of `BaseTestCase`
+
+1. **`field_data`**:
+
+   - This attribute should be overridden in the child class. It contains a list of tuples where each tuple represents a field's name, expected label, expected maximum length, and expected relations count.
+   - Example:
+
+     ```python
+     field_data = [
+         ('name', 'Name', 255, None),
+         ('description', 'Description', 500, None),
+         ('tags', None, None, 3)
+     ]
+     ```
+
+2. **`assertFieldLabel(self, instance, field_name, expected_label)`**:
```

Comment:
This is label if we were using Django views/admin? Suggest adding a note about that, I remember getting confused about it at one point. Also suggest adding a two word field name to show how that translates to a label (eg ('first_name', First Name (?), 255, None)